### PR TITLE
Rework Abins broadening

### LINKED ISSUE #27043: Indirect - Abins: default parameters do not give smooth output
### Expected behavior
With defaults, simple spectra should appear smooth

### Actual behavior
For a calculation of molecular deuterium D2 we get the following plot:

d2-crystal-vib_total-1

Clearly something is going awry during resampling/broadening/rebinning. The phenomenon is greatly reduced, but not eliminated, by increasing the bin width.

### Steps to reproduce the behavior
This behaviour is visible in the high-frequency regime for many Abins spectra, e.g. the squareicn test case.

### Platforms affected
All tried (Mac OSX binary, Ubuntu docker build from source)

### PR DESCRIPTION
**Pull request by ajjackson**
~~This is based on top of #26980 , do not merge until that branch has been merged~~

**Description of work.**
The existing implementation of broadening in Abins has severe problems #27043 

This work replaces the existing implementation with a new system. Several broadening-related functions are provided in a new file AbinsModules/Instruments/Broadening.py ; these are pure functions in an attempt to cut down on the "statefulness" of Abins. It remains the responsibility of an _Instrument_ to know how it should be broadened, and logic is added to dispatch automatically to an efficient broadening function.

The broadening process for these instruments is specialised and cannot use a generic convolution as the broadening kernel is _frequency-dependent_. In order to achieve acceptable performance, a new(?) approach has been developed which interpolates between a limited number of convolved spectra to approximate the results of frequency-dependent broadening. The results are illustrated in #27043 and the method is described briefly in commit messages and docstrings (with ascii art!) It may be worth preparing a technical report or paper on this, depending on how original it turns out to be.

Other methods are also made available: a null function that just resamples the data, a slow but complete summation of Gaussians and an optimised summation of truncated Gaussians which is comparable in performance to the legacy implementation. Variations are also provided that replace the Gaussian with its analytic integral to provide a normal distribution; this is more formally correct than sampling the Gaussian function as our output function is a _histogram_ but in practice the results are very similar and this could be more difficult to compare with reference calculations.

Because this PR fixes the results from Abins, many tests will be broken. Other upcoming changes will also affect results (and tests), so it feels a touch absurd to generate intermediate reference data; @Mantid-Matthew suggested that this PR could target a feature branch in order to permit a series of code reviews with known broken tests. It was decided that in this particular case it is important to get the fixes into master as soon as possible, so I will rebuild the test data and try to get it all working.

**To test:**
Unit tests and system tests have been updated.

For a clear before/after picture, run Abins on one of the sample data files. Observe that the output is now smooth, and previously was not.

If performing before/after testing it would be interesting to track performance differences; the new implementation should be faster than the old one for real-world data.

Fixes #27043 

Release notes have been added

- [x] Add independent tests for new broadening functions
- [x] Generate new test data and eyeball it for reasonableness
- [x] Wait for #26980 to be merged
- [x] Rebase on master and fix all tests
- [x] Write release notes
- [x] Add something to ~~developer~~ docs #27065 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by ajjackson** [comment]
Development notes have been moved into the developer docs section and expanded to describe the interpolated broadening scheme. Another figure is needed before the box can be ticked.

Input is very welcome on where these docs should be located and what they should be named.

**Comment by ajjackson** [comment]
Following offline discussion, the development notes have been moved _back into Concepts_ which seems a better general fit. They are also now a bit more broken-up and less intimidating.

**Comment by ajjackson** [comment]
Seems that everything is working apart from system tests. Which makes sense, because I haven't replaced them yet and the results have changed.

**Comment by ajjackson** [comment]
Before replacing the system test data, here's a comparison plot of the old vs new. The differences look reasonable to me.
Abins-systemtests-comparison

**Comment by ajjackson** [comment]
System tests have passed :tada:  and the Python 3 failure seems to be a spurious MuonQt4 thing.

**Comment by ajjackson** [comment]
There is one thing I would like to examine/address before removing WIP and requesting review.

The assumption made when implementing the interpolated broadening was that the sigma values would be given in ascending order. This applies to TOSCA, but there are instruments in which the energy/broadening relationship would give a descending sigma series - or, worse, a series that goes up _and_ down. The interpolated broadening function should be made robust against this scenario. This _should_ not break the updated integration tests as it will reduce to the current implementation for TOSCA. A test case should check this behaviour directly without needing to instantiate other parts of Abins.

**Comment by ajjackson** [comment]
abins_interp_upanddown
The algorithm now seems to have no trouble tracking multiple turning points in the broadening width (sigma). In this schematic the error appears a bit more more than tolerable at the points where sigma is varying most rapidly. Some consideration should be made of appropriate sigma sampling if/when an instrument is added with a comparably rapidly-varying resolution function.

I think PR is ready for review. I'll add a release note now. I'm sorry that such a large number of files appear to change; this is mostly related to testing and the new content has been limited as far as possible to one new file.

**Comment by ajjackson** [comment]
A couple of areas that I see could be more developed more comprehensively are
- test coverage of the various broadening methods, e.g. against numerical values. Existing tests focus on their approximate correctness against the reference method, so I'm happy with the correctness already. It's a coverage smell thing.
- robustness/helpfulness of the functions with different shapes and types of input. This would be good for re-usability and might clean up the somewhat ugly `broaden_spectrum()` function that dispatches methods and does a lot of axis shuffling. On the other hand, it increases complexity and the number of possible paths through the code without really adding any user-accessible functionality; Abins already knows how to call them correctly.

I'm not convinced those are worth holding up this PR for. There is a possibility that these functions will be refactored into a library with some other parts of Abins (as part of other synergistic activities) so maybe that would be a better time for API-prettifying.

File: scripts/AbinsModules/Instruments/Instrument.py
Comment (Mantid-Matthew) [inline scripts/AbinsModules/Instruments/Instrument.py]: Does this function do anything currently?

**Review by Matthew-Andrew — COMMENTED** [review]
In general looks really good. Level of documentation is particularly good. I have a few minor questions about the code below.

I have not yet done any functional testing. @ajjackson could you provide some simple instructions to get started with this?

File: scripts/AbinsModules/GeneralLoadAbInitioTester.py
Comment (ajjackson) [reply scripts/AbinsModules/GeneralLoadAbInitioTester.py]: Good spot! This was to make the tests remain useful while reworking what advanced parameters exist. Uncommenting it will probably break something and then I'll regenerate the appropriate data files.

File: scripts/AbinsModules/Instruments/Broadening.py
Comment (ajjackson) [reply scripts/AbinsModules/Instruments/Broadening.py]: That does seem more reasonable. Abins in general does use a lot of keywords for essential arguments, I must have missed that somewhere.

Actually `sigma=None` should be allowed anyway, as it isn't used by the null algorithm.

Comment (ajjackson) [reply scripts/AbinsModules/Instruments/Broadening.py]: Going to avoid setting a default that breaks most cases though. Instead will add a docstring note that 'none' ignores sigma, and continue to demand that Sigma is numerical.

Type hinting would be quite handy here, is there much interest in using that in Mantid?

File: scripts/AbinsModules/Instruments/Instrument.py
Comment (ajjackson) [reply scripts/AbinsModules/Instruments/Instrument.py]: It specifies an interface. We could have it `raise NotImplementedError()` to clarify that?

File: scripts/AbinsModules/Instruments/Broadening.py
Comment (Mantid-Matthew) [reply scripts/AbinsModules/Instruments/Broadening.py]: Once we are fully python3 there is quite a lot of interest in doing type hinting and potentially running something like [mypy](https://pypi.org/project/mypy/) to the build servers to check it. I know some people are already using type hinting and just planning to hold off on mergeing the PR's until we are fully python3 which should be soonish.

So in summary if you want to add some type hinting and don't mind waiting a bit before this gets merged I would say go for it.

File: scripts/AbinsModules/Instruments/Instrument.py
Comment (Mantid-Matthew) [reply scripts/AbinsModules/Instruments/Instrument.py]: :+1:  Yes I think having it raise a NotImplimentedError would be good

File: scripts/AbinsModules/Instruments/Broadening.py
Comment (ajjackson) [reply scripts/AbinsModules/Instruments/Broadening.py]: I do mind waiting in this case so will hold off. Once the release is out I might start using it.
One problem is that I believe type hinting is still quite limited for numpy objects.

File: scripts/AbinsModules/GeneralLoadAbInitioTester.py
Comment (ajjackson) [reply scripts/AbinsModules/GeneralLoadAbInitioTester.py]: Have pushed up an update to the test files and restoring this.

However, on reflection I think it does make sense to remove it. Advanced parameters are presumably stored with the reference data in case a parameter modifies how this data is loaded. However, that
- does not happen and
- is not currently planned to happen

so this check could be removed to reduce needless maintenance/replacement of files when the _purpose_ of the test (checking the structure and phonon data import) is unaffected.

**Review by Matthew-Andrew — APPROVED** [review]
Looks good to me :+1:

**Pull request merged**
